### PR TITLE
restructure dataset loading and saving including backup…

### PR DESCRIPTION
**Pull request by Apollon77**
… handling to be more modular for future db usages

### CODE REVIEW

File: packages/db-base/lib/inMemFileDB.js [hunk 1]
```diff
@@ -71,77 +71,84 @@ class InMemoryFileDB {
...
         this.dataDir = this.dataDir.replace(/\\/g, '/');
 
-        // Create data directory
-        if (!fs.existsSync(this.dataDir)) {
-            fs.mkdirSync(this.dataDir);
+        this.datasetName = path.join(this.dataDir, this.settings.fileDB.fileName);
+        const parts = path.dirname(this.datasetName);
+        if (!fs.existsSync(parts)) {
+            fs.ensureDirSync(parts);
         }
```
Comment (AlCalzone) [inline packages/db-base/lib/inMemFileDB.js]: you can get rid of the existsSync - ensurDir does exactly that.

File: packages/db-base/lib/inMemFileDB.js [hunk 2]
```diff
@@ -71,77 +71,84 @@ class InMemoryFileDB {
...
+     * Loads a dataset file
+     *
+     * @param datasetName {string} Filename of the file to load
+     * @returns {object} read data, normally as object
+     */
+    loadDatasetFile(datasetName) {
+        if (!fs.existsSync(datasetName)) {
+            throw new Error(`Database file ${datasetName} does not exists.`);
+        }
+        return fs.readJSONSync(datasetName);
```
Comment (AlCalzone) [inline packages/db-base/lib/inMemFileDB.js:104]: Can we make this async? Should be better for performance if we don't block the event loop until a multi-MB JSON was parsed.
Also least the incremental DB works completely async - or maybe that is not a problem.

File: packages/db-base/lib/inMemFileDB.js [hunk 3]
```diff
@@ -264,50 +271,81 @@ class InMemoryFileDB {
...
+     * Saves the dataset into File incl. handling of a fallback backup file
+     *
+     * @returns {string} JSON string of the complete dataset to also be stored into a compressed backup file
+     */
+    saveDataset() {
         try {
             if (fs.existsSync(this.datasetName)) {
-                const old = fs.readFileSync(this.datasetName);
-                fs.writeFileSync(this.datasetName + '.bak', old);
+                fs.renameSync(this.datasetName, `${this.datasetName}.bak`);
```
Comment (AlCalzone) [inline packages/db-base/lib/inMemFileDB.js]: We should use the overwrite option. I've learned that the hard way (.bak exists -> crash)

File: packages/db-base/lib/inMemFileDB.js [hunk 4]
```diff
@@ -264,50 +271,81 @@ class InMemoryFileDB {
...
+            this.stateTimer = null;
+        }
+    }
+
+    /**
+     * Saves the dataset into File incl. handling of a fallback backup file
+     *
+     * @returns {string} JSON string of the complete dataset to also be stored into a compressed backup file
+     */
+    saveDataset() {
```
Comment (AlCalzone) [inline packages/db-base/lib/inMemFileDB.js]: Also - async save methods maybe?

**Review by AlCalzone — CHANGES_REQUESTED** [review]

File: packages/db-base/lib/inMemFileDB.js (hunk 1 above)
Comment (Apollon77) [reply packages/db-base/lib/inMemFileDB.js]: Ok

File: packages/db-base/lib/inMemFileDB.js (hunk 2 above)
Comment (Apollon77) [reply packages/db-base/lib/inMemFileDB.js:104]: This is done exactly one time at the start of the Main js-Controller. Nothing else will happen before dBs are read and server is there. I see no issue here

File: packages/db-base/lib/inMemFileDB.js (hunk 3 above)
Comment (Apollon77) [reply packages/db-base/lib/inMemFileDB.js]: Kk

File: packages/db-base/lib/inMemFileDB.js (hunk 4 above)
Comment (Apollon77) [reply packages/db-base/lib/inMemFileDB.js]: Only called by Timeouts or on destroy if dB. I do not really see a Benefit if async. Or?!

Comment (Apollon77) [reply packages/db-base/lib/inMemFileDB.js]: Ok, I change to async

Comment (Apollon77) [reply packages/db-base/lib/inMemFileDB.js]: done. the compressed backup is done via a stream and so asny itself ... so I would leave that

**Review by foxriver76 — APPROVED** [review]
lgtm

**Review by AlCalzone — APPROVED** [review]

**Pull request merged**